### PR TITLE
User registration + login.

### PR DESCRIPTION
**Pull request by jeremii**
Please initiate code review process.

### CODE REVIEW

File: backend/sXb-service/sXb-service.csproj [hunk 1]
```diff
@@ -2,13 +2,53 @@
...
+    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.2.3" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="2.2.0" />
+    <PackageReference Include="SendGrid" Version="9.12.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Remove="Extensions\EmailSenderExtensions.cs" />
```
Comment (jeremii) [inline backend/sXb-service/sXb-service.csproj]: Needs deleted

File: backend/sXb-service/Controllers/UsersController.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            if (item.Email != null)
+            {
+                user.Email = item.Email;
+            }
+            Repo.Update(user);
+
+            return RedirectToAction("GetAll");
+        }
+
+        //TODO: Create Viewmodel for user password, don't pass in url.
```
Comment (LeviButcher) [inline backend/sXb-service/Controllers/UsersController.cs]: Don't add TODO to code. They get out of date like this one.

File: backend/sXb-service/Controllers/UsersController.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                var code = await _userManager.GenerateEmailConfirmationTokenAsync(user);
+                var callbackUrl = Url.Page(
+                    "/emailconfirmed",
+                    pageHandler: null,
+                    values: new { userId = user.Id, code = code },
+                    protocol: Request.Scheme);
+                await _emailSender.SendEmailAsync(user.Email, "Confirm your email",
+                    $"Please confirm your account by <a href='{HtmlEncoder.Default.Encode(callbackUrl)}'>clicking here</a>.");
+
+                // Uncomment for registration w/o email confirmation.
```
Comment (LeviButcher) [inline backend/sXb-service/Controllers/UsersController.cs]: Remove comments

File: backend/sXb-service/notes.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# SQL Server setup (Linux)
+
+docker run -e 'ACCEPT_EULA=Y' -e 'MSSQL_SA_PASSWORD=DevPassword1' -p 1433:1433 --name sxbdb -d microsoft/mssql-server-linux
```
Comment (jeremii) [inline backend/sXb-service/notes.md]: I'll move this file to local device.

File: backend/sXb-service/EF/DataInitializer.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            SeedData(context);
+        }
+     
+        public static void ClearData(Context context)
+        {
+            
+            DeleteRowsFromTable(context, "dbo", "AspNetUsers");
+            
+        }
+        public static void ResetAllIdentities (Context context)
```
Comment (LeviButcher) [inline backend/sXb-service/EF/DataInitializer.cs]: Remove Empty Methods unless absolutely necessary.

File: backend/sXb-service/Models/Identity/AccountViewModels/LoginWith2faViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace sXb_service.Models.Identity.AccountViewModels
+{
```
Comment (LeviButcher) [inline backend/sXb-service/Models/Identity/AccountViewModels/LoginWith2faViewModel.cs]: We aren't using 2FA, this should be deleted.

File: backend/sXb-service/SampleData/SampleData.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                FirstName = first, LastName = last,
+                Email = first.Substring(0, 1).ToLower() + last.ToLower() + emailSuffix,
+                NormalizedEmail = first.Substring(0, 1).ToUpper() + last.ToUpper() + emailSuffix.ToUpper(),
+                UserName = first.Substring(0,1).ToLower() + last.ToLower(),
+                NormalizedUserName = first.Substring(0, 1).ToUpper() + last.ToUpper(),
+                EmailConfirmed = true
+            };
+            user.PicturePath = "random guy.jpeg";
+            user.PasswordHash = new PasswordHasher<User>().HashPassword(user, "Test#1");
+            //user.RSSFeeds = (List<Rss>)GetRssFeeds(user);
```
Comment (LeviButcher) [inline backend/sXb-service/SampleData/SampleData.cs]: Delete

File: backend/sXb-service/notes.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# SQL Server setup (Linux)
+
```
Comment (LeviButcher) [inline backend/sXb-service/notes.md]: This should be in a wiki, not within codebase.

File: frontend/notes.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Setup
+
```
Comment (LeviButcher) [inline frontend/notes.md]: Should be in wiki.

File: frontend/package.json [hunk 10]
```diff
@@ -3,17 +3,44 @@
...
+    "eslint-config-airbnb": "^17.1.0",
+    "eslint-config-prettier": "^4.3.0",
+    "eslint-plugin-prettier": "^3.1.0",
+    "lint-staged": "^8.2.1",
+    "prettier": "^1.18.2",
+    "pretty-quick": "^1.11.1"
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "BROWSER=firefox react-scripts start",
```
Comment (LeviButcher) [inline frontend/package.json]: Why is there a BROWSER=firefox here?

File: frontend/src/pages/Register.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      <h1>Register</h1>
+
+      <form
+        onSubmit={e => {
+          e.preventDefault();
+          e.persist();
+          ApiPost("users/new", false, { username, email, password }).then(
+            res =>
+              // After registering, show verify email page.
+              (window.location.href = "/verify-email?email=" + email)
```
Comment (LeviButcher) [inline frontend/src/pages/Register.js]: reach router has a navigate function. I believe using window.location.href will cause a full refresh of react components so everything gets mounted again. Using navigate will only unmount and mount the required components.

**Pull request merged**